### ISSUE TITLE
Ensure only schemas in `db-schemas` and `db-extra-search-path` are scanned for the schema cache

### ISSUE THREAD

**Issue by steve-chavez** [PostgREST/postgrest#2015]
Some DbStructure queries are not yet filtered by the schemas on `db-schemas` and `db-extra-search-path`:

- `allTables`:

https://github.com/PostgREST/postgrest/blob/5cd7d35966b8e8481215ca181c7629191e7e470e/src/PostgREST/DbStructure.hs#L488

- `allM2ORels`

- `allPrimaryKeys`:
 
https://github.com/PostgREST/postgrest/blob/5cd7d35966b8e8481215ca181c7629191e7e470e/src/PostgREST/DbStructure.hs#L736

Taken from https://github.com/PostgREST/postgrest/issues/2013#issuecomment-962438926.

**Comment by steve-chavez** [comment]
Right now the [db-extra-search-path docs](https://postgrest.org/en/v8.0/configuration.html#db-extra-search-path) say:

> This parameter was meant to make it easier to use PostgreSQL extensions (like PostGIS) that are outside of the db-schema.

Schemas where extensions are installed(`public` or `extensions`) can contain many tables and that conflicts with the purpose of not scanning schemas unnecessarily.

Maybe there should be another config param.

**Comment by wolfgangwalther** [comment]
I don't think we need to scan all the tables in `db-extra-search-path` by default. Right now, we're starting with scanning all tables in the whole database. The change should not be to just limit this to the tables in the public and extra schemas - but to reverse the whole process:
- first query for all views in the public schema and find out their base columns / tables
- then query for all tables in the public schema + the hidden base tables
- then columns, etc...

Of course, this is most efficiently done in a single query - think #1612.

**Comment by steve-chavez** [comment]
> but to reverse the whole process

That's a great idea, I think it would also cover the use case where only functions are exposed:

> Some users only expose functions through PostgREST so they don't really need embedding, that's where the none option could be useful(detecting relationships is generally fast, but could still take a while on big databases).

(From https://github.com/PostgREST/postgrest/issues/1783#issuecomment-962370105)

Since no table/view resource is exposed then functions wouldn't be able to embed(they don't have a target). In this case we could omit relationship scanning.

**Comment by yilinjuang** [comment]
Some real world data for this. 

We run schema-per-tenant. About 500 schemas in the DB, only about 10% of them are exposed to PostgREST.
The fk query `allM2OandO2ORels` scans all schemas and took 15.2 seconds (425k rows). When the query is scoped to db-schemas and db-extra-search-path, it took 1 second (26k rows). 

`allComputedRels` can benefit from scoping to exposed schemas too.
